File: tareas/2/24190112/17_Reverse_polish_notation/Reverse_polish_notation.py
#Demasiadas funciones para poder escribir correctamente en el archivo de salida
def comando_de_ayuda():

    return ("Ingrese expresiones en RPN,\n"
            "en donde los operadores aparecen luego de\n"
            "los operandos sobre los cuales serán aplicados.\n"
            "Cada línea consiste de número, un operador,\n"
            "o uno de los siguientes comandos:\n"
            "Q: Terminar el programa\n"
            "H: Mostrar este mensaje de ayuda\n"
            "C: Limpiar la memoria de la calculadora\n")

def es_operador(s):

    return s == '+' or s == '-' or s == '*' or s == '/'

def aplicar_operador(operador, memoria):

    ultimo = memoria.pop()
    penultimo = memoria.pop()

    if operador == "+":
        resultado = penultimo + ultimo
    elif operador == "-":
        resultado = penultimo - ultimo
    elif operador == "*":
        resultado = penultimo * ultimo
    elif operador == "/":
        resultado = penultimo / ultimo

    memoria.append(resultado)

    return resultado

def main():
    archivo_entrada = "entrada_ejercicio17.txt"
    archivo_salida = "salida_ejercicio17.txt"

    with open(archivo_entrada, encoding="utf-8") as entrada:
        entrada = entrada.readlines()
    with open(archivo_salida, 'w', encoding="utf-8") as salida:
            salida.write(comando_de_ayuda() + "\n\n")

            memoria = []
            #Lista vacía

            for linea in entrada:
                entrada = linea.strip()
                #Para eliminar espacios en blanco

                if entrada == "Q":
                    break

                elif entrada == "C":
                    memoria.clear()

                elif entrada == "H":
                    salida.write(comando_de_ayuda() + "\n\n")

                elif es_operador(entrada):
                    resultado = aplicar_operador(entrada, memoria)

                else:
                        memoria.append(float(entrada))

            salida.write(f"Resultado final: {memoria[-1]}\n")
            #Con los numeros negativos accedemos desde el ultimo indice de la lista
                    

if __name__ == "__main__":
    main()

File: ejercicios/19_20_21_22/reverse_polish_notation/reverse_polish_notation.cpp
#include <iostream>
#include <stack>
#include <string>
using namespace std;

void ComandoDeAyuda(void) {
    cout << "Ingrese expresiones en RPN," << endl;
    cout << "en donde los operadores aparecen luego de" << endl;
    cout << "los operandos sobre los cuales serán aplicados." << endl;
    cout << "Cada línea consiste de número,";
    cout << " un operador,";
    cout << " o uno de los siguientes comandos:" << endl;
    cout << "Q: Terminar el programa" << endl;
    cout << "H: Mostrar este mensaje de ayuda" << endl;
    cout << "C: Limpiar la memoria de la calculadora" << endl;
}

bool EsOperador(string s) {
    return s == "+" || s == "-" || s == "*" || s == "/";
}

void AplicarOperador(string operador, stack<double>& memoria) {
    double resultado;
    double derecha = memoria.top();
    memoria.pop();
    double izquierda = memoria.top();
    memoria.pop();
    if (operador == "+") {
        resultado = izquierda + derecha;
    } else if (operador == "-") {
        resultado = izquierda - derecha;
    } else if (operador == "*") {
        resultado = izquierda * derecha;
    } else if (operador == "/") {
        resultado = izquierda / derecha;
    }
    cout << resultado << endl;
    memoria.push(resultado);
}

int main(void) {
    cout << "Simulación de calculadora RPN (ingrese H para ayuda)" << endl;
    stack<double> memoria;
    while (true) {
        string entrada;
        cin >> entrada;
        if (entrada == "Q") {
            break;
        } else if (entrada == "C") {
            memoria = stack<double>();
        } else if (entrada == "H") {
            ComandoDeAyuda();
        } else if (EsOperador(entrada)) {
            AplicarOperador(entrada, memoria);
        } else {
            memoria.push(stod(entrada));
        }
    }
    return 0;
}